### ISSUE TITLE
有关loss的问题

### ISSUE THREAD

**Issue by bobo0810** [xialuxi/CircleLoss_Face#5]
@xialuxi  非常感谢你的工作！同时我存在一些疑问：

1、在[https://github.com/TinyZeaMays/CircleLoss/blob/master/circle_loss.py](url)  的第37、39行与本文有所差异，可以解释一下原因吗?
`dim=1 or  dim=0`  和    `loss or   loss.mean()`   

2、我试着用你的代码跑人证识别，75人的200多万照片，class_num 为75万。参数设置为 margin=0.25  gamma=256 or 128，batch_size=256, 训练过程中损失非常大，并且无法收敛。麻烦问一下可能是什么原因呢？

**Comment by xialuxi** [comment]
一个是样本的维度，一个是类别的维度，我自己训练ms1m的数据是正常的，方便的话，贴你下你的训练日志，或者你怎么调用的？

**Comment by bobo0810** [comment]
> 一个是样本的维度，一个是类别的维度，我自己训练ms1m的数据是正常的，方便的话，贴你下你的训练日志，或者你怎么调用的？

非常感谢你的帮助！我是基于 标签训练，self.weight是我的类别权重矩阵。设置dim=1和loss,mean()，与本文一致。

![5c8e91814c7eef0a9acf21e6350c627](https://user-images.githubusercontent.com/26057879/88289927-90cd9700-cd28-11ea-9288-7161a1a0397f.png)

训练代码如下
` losses, correct = classifier_model(feature, target)`
`   loss = losses / iter_size`
` loss.backward()`

训练结果   由于GPU限制，我是batch=32, 前向8次再更新，曲线实现理论batch=256。训练较慢，每隔100个理论batch打印一条信息。看到损失很大并且不再下降，我就停止训练了。希望能看下是哪儿出了问题？会是margin、gamma参数的问题吗？
![81e6cf01e36317411e79894b54651fc](https://user-images.githubusercontent.com/26057879/88290534-82cc4600-cd29-11ea-94a6-523cc752db90.png)

**Comment by xialuxi** [comment]
迭代多次累积梯度，再反向传播，你的代码有问题，再就是你的学习率过大，建议使用warmup。

**Comment by bobo0810** [comment]
> 迭代多次累积梯度，再反向传播，你的代码有问题，再就是你的学习率过大，建议使用warmup。

非常感谢您的建议，我准备尝试warmup。
1、分类器权重为[75万，128维] ，您建议怎么设置gamma,margin参数？
2、分类器权重 加载 AMSoftmax (scale=30,margin=0.4)的预训练参数, CircleLoss设定scale=128/256，margin=0.25。理论batch=256训练，损失从150多开始不收敛，batch准确率为0/256，猜测因为sacle,margin不匹配导致类似重新训练？
3、之后将CircleLoss改为scale=30，margin=0.4，分类器权重仍加载AMSoftmax，损失从33开始下降，batch准确率一直在255/256左右，训练过程中学习率由0.1改为0.01，目前降到4.7左右。是否正常？
4、我补充一下梯度更新代码，这样写应该没问题吧？
![3745b54586d2123083313a1dd936eab](https://user-images.githubusercontent.com/26057879/88358381-4b00e500-cda1-11ea-99ac-a04b94065c92.png)
最后，再次感谢您的帮助，对我帮助很大！

**Comment by xialuxi** [comment]
loss不要除以iter_size

**Comment by xialuxi** [comment]
gamma,margin参数最好参考论文更改，而不是直接使用amsoftmax的参数，由于你的训练id较多，收敛慢和困难都是正常现象，梯度累加代码没有问题,只是iter_size太大，会影响bn层的参数，太小，则收敛更加困难．另外这个损失函数的初始loss比较大，多训练loss会降下来，建议使用余弦退火的学习率更新策略．

**Comment by bobo0810** [comment]
> gamma,margin参数最好参考论文更改，而不是直接使用amsoftmax的参数，由于你的训练id较多，收敛慢和困难都是正常现象，梯度累加代码没有问题,只是iter_size太大，会影响bn层的参数，太小，则收敛更加困难．另外这个损失函数的初始loss比较大，多训练loss会降下来，建议使用余弦退火的学习率更新策略．

非常感谢，我想再问下  假设epoch=100, 学习率=0.1，经验上CosineAnnealingLR应该如何设置？
（1）设定T_max=epoch=100，以实现学习率变化0.1=>0  
（2）T_max应该设小些，以实现学习率 先下降后上升的反复循环
![595bc1374eb4b16b04f9e3af9854df5](https://user-images.githubusercontent.com/26057879/88747832-2b433580-d182-11ea-8534-bc8228e8407a.png)

**Comment by bobo0810** [comment]
> gamma,margin参数最好参考论文更改，而不是直接使用amsoftmax的参数，由于你的训练id较多，收敛慢和困难都是正常现象，梯度累加代码没有问题,只是iter_size太大，会影响bn层的参数，太小，则收敛更加困难．另外这个损失函数的初始loss比较大，多训练loss会降下来，建议使用余弦退火的学习率更新策略．

非常感谢，我想再问下:
1、余弦退火策略 应用在整个训练过程中吗？还是仅前几轮warm-up呢？
2、假设epoch=100, 学习率=0.1，经验上CosineAnnealingLR应该如何设置？
（1）设定T_max=epoch=100，以实现学习率变化0.1->0
（2）T_max应该设小些，以实现学习率 先下降后上升的反复循环
![595bc1374eb4b16b04f9e3af9854df5](https://user-images.githubusercontent.com/26057879/88748718-4b73f400-d184-11ea-879c-82205ebe0cfc.png)
